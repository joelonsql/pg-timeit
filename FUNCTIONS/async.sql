CREATE OR REPLACE FUNCTION pit.async(
    function_name text,
    input_values text[] DEFAULT ARRAY[]::text[],
    significant_figures integer DEFAULT 1
)
RETURNS bigint
LANGUAGE plpgsql
AS $$
<<fn>>
declare
    id bigint;
begin

    if num_nulls(function_name,input_values,significant_figures) <> 0
    then
        raise exception 'no arguments must be null';
    end if;

    INSERT INTO pit.tests
        (test_state)
    VALUES
        ('init')
    RETURNING tests.id INTO id;

    INSERT INTO pit.test_params
        (id, function_name, input_values, significant_figures)
    VALUES
        (id, function_name, input_values, significant_figures);

    return id;

end
$$;

CREATE OR REPLACE FUNCTION pit.async(
    function_name text,
    significant_figures integer
)
RETURNS numeric
LANGUAGE sql
AS $$
SELECT pit.async($1,ARRAY[]::text[],$2);
$$;
