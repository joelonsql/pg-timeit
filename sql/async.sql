CREATE EXTENSION pit;

--
-- Enqueue new test.
--
SELECT pit.async('pg_sleep',ARRAY['0.1']);

--
-- Work until there is no more work.
--
-- Ignore notice messages since we can't know how
-- many iterations that will be necessary.
--
SET client_min_messages TO 'warning';
CALL pit.work(return_when_idle := true);

--
-- Have a look at test result.
--
SELECT
    id,
    test_state,
    executions,
    final_result
FROM pit.tests;

DROP EXTENSION pit;
