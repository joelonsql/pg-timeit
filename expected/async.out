CREATE EXTENSION pit;
--
-- Enqueue new test.
--
SELECT pit.async('pg_sleep',ARRAY['0.1']);
 async 
-------
     1
(1 row)

--
-- Work until there is no more work.
--
-- Ignore notice messages since we can't know how
-- many iterations that will be necessary.
--
SET client_min_messages TO 'warning';
CALL pit.work(return_when_idle := true);
--
-- Have a look at test result.
--
SELECT
    id,
    test_state,
    executions,
    final_result
FROM pit.tests;
 id | test_state | executions | final_result 
----+------------+------------+--------------
  1 | final      |          1 |          0.1
(1 row)

DROP EXTENSION pit;
